File: solutions/elixir/pascals-triangle/lib/pascals_triangle.ex
defmodule PascalsTriangle do
  @doc """
  Calculates the rows of a pascal triangle
  with the given height
  """
  @spec rows(integer) :: [[integer]]
  def rows(n) do
    for r <- 1..n, do: row(r)
  end 

  defp row(1), do: [1]
  defp row(n) do
    prev_row = row(n - 1)
    Enum.zip_with([0 | prev_row], prev_row ++ [0], &+/2)
  end
end

File: solutions/julia/pascals-triangle/pascals-triangle.jl
function triangle(maxN)
    if maxN < 0
        throw(DomainError(maxN, "Pascal's triangle undefined for negative rows"))
    end
    
    allRows = Vector{Int64}[]
    for n in 1:maxN
        row = Int64[]
        for k in 1:n
            push!(row, binomial(n - 1, k - 1))
        end
        push!(allRows, row)
    end
    allRows
end